eventbus: add metrics

sukunrt opened the pull request:
Related to: https://github.com/libp2p/go-libp2p/issues/2020

sukunrt commented:
You'll need to change 
```
"datasource": {
                                "type": "prometheus",
                                "uid": "aMik1H0Vz"
                        },
                        
  ```
datasource.uid to your prometheus datasource uid.

marten-seemann commented on p2p/host/eventbus/basic_metrics.go:
```diff
@@ -0,0 +1,44 @@
+}
+
+func (m *metricsTracer) EventEmitted(typ reflect.Type) {
+	eventsEmitted.WithLabelValues(typ.String()).Inc()
```
Should we filter out the package name here (the `event.`)?

This is how it looks like in Grafana:
![image](https://user-images.githubusercontent.com/1478487/216177149-4d591567-1749-4508-b663-3f54402d7c28.png)

sukunrt replied:
fixed.

marten-seemann commented on p2p/host/eventbus/basic_metrics.go:
```diff
@@ -0,0 +1,147 @@
+		},
+		[]string{"event"},
+	)
+	subscriberQueueLength = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: metricNamespace,
+			Name:      "subscriber_queue_length",
+			Help:      "Subscriber queue length",
+			Buckets:   prometheus.ExponentialBuckets(1.0, 2.0, 10),
+		},
+		[]string{"subscriber_name"},
+	)
+	subscriberQueueFull = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: metricNamespace,
+			Name:      "subscriber_queue_full",
+			Help:      "Subscriber Queue completely full",
+		},
+		[]string{"subscriber_name"},
+	)
+	subscriberEventQueued = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: metricNamespace,
+			Name:      "subscriber_event_queued",
+			Help:      "Event Queued for subscriber",
+		},
+		[]string{"subscriber_name"},
+	)
```
I'd like to avoid having subscribers pass in a name. Can we just make this per event?

sukunrt replied:
But the queues are per subscriber and not per event. 
We can remove the subscriber metrics and just track whether for a given event type one of the subscriber queues was full.

marten-seemann commented on p2p/host/eventbus/basic.go:
```diff
@@ -57,12 +67,22 @@ func NewBus() event.Bus {
 	}
 }
 
+func NewBusWithMetrics() event.Bus {
```
This works totally fine, but we usually use the option pattern:
```go
type Option func(*basicBus) error

func NewBus(opts ...Option) event.Bus
```

sukunrt replied:
will fix. This would also be backwards compatible. nice :)

marten-seemann replied:
This is what we did for the swarm: https://github.com/libp2p/go-libp2p/blob/master/p2p/net/swarm/swarm.go#L74-L79

sukunrt replied:
fixed. Not returning error for now since there's only one option that can't error.

marten-seemann commented on p2p/host/eventbus/basic_metrics.go:
```diff
@@ -0,0 +1,147 @@
+	subscriberEventQueued.WithLabelValues(name).Inc()
+}
+
+func stripPrefix(s string) string {
```
Nit: https://pkg.go.dev/strings#TrimPrefix

sukunrt replied:
fixed.

sukunrt commented:
@marten-seemann Addressed review comments. Removed custom passed in names from all subscribers for now. they are set to the filename. If you want I'll add them back.

marten-seemann commented on p2p/host/eventbus/basic_metrics.go:
```diff
@@ -0,0 +1,139 @@
+	notificationTime = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: metricNamespace,
+			Name:      "notification_time_microseconds",
```
There's a convention in Prometheus to use SI units: https://prometheus.io/docs/practices/naming/. So this should be `notification_time_seconds`.

marten-seemann replied:
Thinking about this a bit more, this measurement doesn't seem as useful any more:
* If all channels don't block, this takes basically no time at all, we're just adding an event to every channel.
* If a channel blocks because the subscriber is not reading from the channel, it will block indefinitely, and then this event won't be emitted.

I suggest remote it.

sukunrt replied:
fixed

marten-seemann commented on p2p/host/eventbus/basic_metrics.go:
```diff
@@ -0,0 +1,139 @@
+	numSubscribers = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: metricNamespace,
+			Name:      "num_subscribers",
```
```suggestion
			Name:      "subscribers_total",
```

marten-seemann replied:
In Grafana, this would look pretty cool as a Gauge: 
<img width="1365" alt="image" src="https://user-images.githubusercontent.com/1478487/216519035-4fa60f3f-ea66-4d5d-8829-1eded5ce735b.png">

sukunrt replied:
fixed

marten-seemann commented on p2p/host/eventbus/grafana-dashboards/eventbus.json:
```diff
@@ -0,0 +1,835 @@
+            "uid": "${DS_PROMETHEUS}"
+          },
+          "editorMode": "builder",
+          "expr": "sum by(event) (rate(libp2p_eventbus_events_emitted_total[1m]))",
```
It's better to use the `$__rate_interval` variable instead of hardcoding a value. Also, I don't think you need the `sum`. This configuration should work:
<img width="919" alt="image" src="https://user-images.githubusercontent.com/1478487/216519715-8ab80908-0430-4865-9ccf-7a684e9a33de.png">

sukunrt replied:
removed hardcoding

sukunrt commented:
@marten-seemann  addressed review comments.

sukunrt commented:
made subscriber queue a state timeline

MarcoPolo commented on p2p/host/eventbus/basic_metrics.go:
```diff
@@ -0,0 +1,121 @@
+		},
+		[]string{"event"},
+	)
+	subscriberQueueLength = prometheus.NewHistogramVec(
```
Why is this a histogram? I think this should be a gauge. What do you think was the value in having this be a histogram?

If it's a gauge, the promql query also becomes simpler and correct. I think the current query in the grafana dashboard is wrong. It's showing you the increase in queue length. If you're full, you'll see that number go to 0 since it's not increasing.

sukunrt replied:
I wanted to look at things like 90%ile queue length. I guess we can ignore that since if the consumer is slow the queue will get full eventually.  Having a gauge alerts us to increasing queue length which should be enough. 

I didn't realise the current query would go to 0, I thought sum / count would stay constant. I just checked, it does go to 0 😅

sukunrt commented on p2p/host/eventbus/basic_metrics.go:
```diff
@@ -0,0 +1,121 @@
+var initMetricsOnce sync.Once
+
+func initMetrics() {
+	prometheus.MustRegister(eventsEmitted, totalSubscribers, subscriberQueueLength, subscriberQueueFull, subscriberEventQueued)
```
fixed. I'm using 
```
type MetricsTracerOption = func(*metricsTracerSetting)

func NewMetricsTracer(opts ...MetricsTracerOption) MetricsTracer
```

marten-seemann reviewed (approved):
I added two minor changes to this PR:
1. Removing the version number from the subscriber event name. This makes it more pleasant to view in Grafana.
2. A number of minor tweaks to the Grafana dashboard.

I think this is ready to go now. Thanks to @sukunrt for doing all the heavy lifting here!
